lib/Sema/CSSimplify.cpp

```diff
@@ -1410,6 +1410,35 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
     case ConstraintKind::ArgumentConversion:
     case ConstraintKind::OperatorArgumentTupleConversion:
     case ConstraintKind::OperatorArgumentConversion:
+      if (type1->is<InOutType>() && !type2->is<InOutType>() && typeVar2) {
+        // Try to upgrade type2 to an InOutType
+        auto inOutType1 = type1->getAs<InOutType>();
+        auto typeVar3 = createTypeVariable(getConstraintLocator(locator),
+                                           typeVar2->getImpl().getOptions());
+        auto inOutType3 = InOutType::get(typeVar3);
+        
+        TypeMatchOptions subFlags = flags;
+        subFlags -= TMF_GenerateConstraints;
+        // First, see if we can actually replace typeVar2 with an InOutType.
```

Comment:
`matchTypes()` isn't permitted to tentatively do testing this way... when it fails, that branch of the solution space is invalid. In cases where it isn't obvious which way to go, we end up having to introduce a disjunction so the solver itself can explore both paths (and use the scope RAII objects to allow some of them to fail). I wonder if you could tackle this by adding a `TVO_CanBeInOut` flag to specific type variables, and only allow type variables with that flag set to bind to `inout` types, similarly to the way we have the `TVO_MustBeMaterializable` flag.